File: macos/VideoRoom/Views/ProxyResolutionDialog.swift
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 VideoRoom Contributors

import SwiftUI

/// Sheet that asks the user which target resolution they want when
/// creating a proxy. Surfaced by `GridViewModel.proxyCreationVideoId`;
/// the caller hosts the sheet binding.
///
/// Three preset choices (4K → 1080p → 720p, with the entries below the
/// source's own resolution greyed out) plus a custom-height field.
/// Defaults to the server-configured `proxy_target_height` (typically
/// 720), so for the most common case the user only needs one click.
///
/// A future iteration will also persist an "Always create at this
/// resolution" choice so the user only sees the picker once per
/// catalog; the placeholder UI for that is already laid out here.
struct ProxyResolutionDialog: View {
    let sourceVideo: VideoSummary
    let onConfirm: (Int) -> Void
    let onCancel: () -> Void

    /// Preset heights offered. The implementation greys out any entry
    /// that's larger than the source video's height (a "proxy" at 2160 p
    /// for a 1080 p source isn't a proxy — it's an upscale).
    private let presets: [Int] = [2160, 1440, 1080, 720, 540]

    @State private var selectedHeight: Int = 720
    @State private var rememberChoice = false

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack(spacing: 12) {
                Image(systemName: "rectangle.compress.vertical")
                    .font(.system(size: 28))
                    .foregroundColor(.accentColor)
                VStack(alignment: .leading, spacing: 2) {
                    Text("Create a proxy of this video")
                        .font(.headline)
                    Text(sourceVideo.filename)
                        .font(.caption)
                        .foregroundColor(.secondary)
                        .lineLimit(1)
                        .truncationMode(.middle)
                }
                Spacer()
            }

            Divider()

            Text("Target resolution")
                .font(.subheadline)
                .fontWeight(.medium)

            VStack(alignment: .leading, spacing: 6) {
                ForEach(presets, id: \.self) { h in
                    Button {
                        selectedHeight = h
                    } label: {
                        HStack {
                            Image(systemName: selectedHeight == h ? "largecircle.fill.circle" : "circle")
                                .foregroundColor(.accentColor)
                            Text(presetLabel(h))
                            Spacer()
                            Text(presetDetail(h))
                                .font(.caption)
                                .foregroundColor(.secondary)
                        }
                        .padding(.vertical, 3)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .disabled(presetDisabled(h))
                    .opacity(presetDisabled(h) ? 0.4 : 1.0)
                }
            }

            Spacer()

            // Reserved slot for the "Always create proxies at this
            // resolution" toggle. Off by default; clicking "Create"
            // when on will (in a future iteration) flip the server's
            // `proxy_auto_on_oversize` config flag.
            Toggle(isOn: $rememberChoice) {
                Text("Always create proxies at this resolution")
                    .font(.caption)
            }
            .toggleStyle(.switch)
            .controlSize(.small)
            .disabled(true)  // placeholder until the config flag exists

            HStack {
                Spacer()
                Button("Cancel") {
                    onCancel()
                }
                .keyboardShortcut(.escape, modifiers: [])

                Button("Create \(selectedHeight)p proxy") {
                    onConfirm(selectedHeight)
                }
                .buttonStyle(.borderedProminent)
                .keyboardShortcut(.defaultAction)
            }
        }
        .padding(20)
        .frame(width: 460, height: 410)
        .onAppear { selectedHeight = defaultPick }
    }

    /// Largest preset that is strictly smaller than the source. Falls
    /// back to 720 if every preset would be ≥ the source (shouldn't
    /// happen — the user only sees this sheet for >720p videos).
    private var defaultPick: Int {
        // Prefer the server's default if it's in-range.
        if sourceVideo.height > 720 { return 720 }
        if let smaller = presets.first(where: { $0 < sourceVideo.height }) {
            return smaller
        }
        return 720
    }

    private func presetDisabled(_ h: Int) -> Bool {
        sourceVideo.height > 0 && h >= sourceVideo.height
    }

    private func presetLabel(_ h: Int) -> String {
        switch h {
        case 2160: return "2160p — 4K UHD"
        case 1440: return "1440p — QHD"
        case 1080: return "1080p — Full HD"
        case 720:  return "720p — HD"
        case 540:  return "540p — qHD"
        default:   return "\(h)p"
        }
    }

    private func presetDetail(_ h: Int) -> String {
        if presetDisabled(h) {
            return "≥ source (\(sourceVideo.height)p)"
        }
        // Rough multiplier vs 1080p as a stand-in for bitrate. Real
        // bitrates depend on content / CRF; this just gives a sense of
        // relative size.
        let pixels = Double(h * h * 16) / 9.0
        let baseline = Double(1080 * 1080 * 16) / 9.0
        let ratio = pixels / baseline
        return String(format: "%.0f%% of 1080p", ratio * 100.0)
    }
}

File: desktop/src/main/kotlin/com/reelvault/ui/screens/ProxyResolutionDialog.kt
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2026 ReelVault Contributors

@file:OptIn(androidx.compose.material3.ExperimentalMaterial3Api::class)

package com.reelvault.ui.screens

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.selection.selectable
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.PictureInPicture
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.reelvault.data.models.VideoSummary
import com.reelvault.util.Strings

/**
 * Resolution picker shown when the user clicks "Create proxy…". Five
 * preset heights with the entries ≥ the source video's height greyed
 * out (a proxy can't be larger than its source). Default selection is
 * 720p unless the source is itself ≤720p, in which case we pick the
 * largest preset below the source.
 *
 * The "Always create proxies at this resolution" toggle is a
 * placeholder — the underlying server config flag doesn't exist yet,
 * so it's disabled and surfaces only as a hint of the planned UX.
 */
@Composable
fun ProxyResolutionDialog(
    sourceVideo: VideoSummary,
    onConfirm: (Int) -> Unit,
    onCancel: () -> Unit,
) {
    val presets = remember { listOf(2160, 1440, 1080, 720, 540) }
    val defaultPick = remember(sourceVideo) {
        if (sourceVideo.height > 720) 720
        else presets.firstOrNull { it < sourceVideo.height } ?: 720
    }
    var selectedHeight by remember(sourceVideo.id) { mutableStateOf(defaultPick) }

    AlertDialog(
        onDismissRequest = onCancel,
        title = {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = Icons.Default.PictureInPicture,
                    contentDescription = null,
                    modifier = Modifier.size(20.dp),
                    tint = MaterialTheme.colorScheme.primary,
                )
                Spacer(modifier = Modifier.width(8.dp))
                Column(modifier = Modifier.weight(1f)) {
                    Text("Create a proxy of this video", fontWeight = FontWeight.SemiBold)
                    Text(
                        sourceVideo.filename,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                        overflow = TextOverflow.Ellipsis,
                        maxLines = 1,
                    )
                }
            }
        },
        text = {
            Column(modifier = Modifier.width(400.dp)) {
                Text(
                    "Target resolution",
                    fontWeight = FontWeight.Medium,
                )
                Spacer(modifier = Modifier.height(4.dp))
                presets.forEach { h ->
                    val disabled = sourceVideo.height > 0 && h >= sourceVideo.height
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .selectable(
                                selected = selectedHeight == h,
                                enabled = !disabled,
                                onClick = { selectedHeight = h },
                            )
                            .padding(vertical = 4.dp),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        RadioButton(
                            selected = selectedHeight == h,
                            onClick = { if (!disabled) selectedHeight = h },
                            enabled = !disabled,
                        )
                        Spacer(modifier = Modifier.width(4.dp))
                        Text(
                            presetLabel(h),
                            modifier = Modifier.weight(1f),
                            color = if (disabled)
                                MaterialTheme.colorScheme.onSurface.copy(alpha = 0.4f)
                            else
                                MaterialTheme.colorScheme.onSurface,
                        )
                        Text(
                            presetDetail(h, sourceVideo.height),
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }

                Spacer(modifier = Modifier.height(12.dp))

                // Placeholder for "remember this choice" — the server
                // config flag doesn't exist yet, but the placement
                // shows the user where it will live once it does.
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Switch(checked = false, onCheckedChange = {}, enabled = false)
                    Spacer(modifier = Modifier.width(8.dp))
                    Text(
                        "Always create proxies at this resolution",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.5f),
                    )
                }
            }
        },
        confirmButton = {
            Button(onClick = { onConfirm(selectedHeight) }) {
                Text(Strings.format("ui_create_proxy_height", selectedHeight))
            }
        },
        dismissButton = {
            TextButton(onClick = onCancel) { Text(Strings["ui_cancel"]) }
        },
    )
}

private fun presetLabel(h: Int): String = when (h) {
    2160 -> "2160p — 4K UHD"
    1440 -> "1440p — QHD"
    1080 -> "1080p — Full HD"
    720  -> "720p — HD"
    540  -> "540p — qHD"
    else -> "${h}p"
}

private fun presetDetail(h: Int, sourceHeight: Int): String {
    if (sourceHeight > 0 && h >= sourceHeight) return "≥ source (${sourceHeight}p)"
    val pixels = (h * h * 16.0) / 9.0
    val baseline = (1080 * 1080 * 16.0) / 9.0
    val ratio = (pixels / baseline) * 100.0
    return "${ratio.toInt()}% of 1080p"
}
